### PR TITLE
Show max scores for assignments instead of final submission scores

### LINKED ISSUE #906: Students should be able to see scores that are not on their final submission
Currently, we only show scores for their final submission. Instead, we should show the score tag with the max score for each score kind.

### PR DESCRIPTION
**Pull request by knrafto**
Resolves #906 

Log in as nikkid@gmail.com and publish "total" and "correctness" scores.

You can see all scores for hog here: http://localhost:5000/admin/course/1/nikkid%40gmail.com/2

At these pages, you should see the max scores. Scores will link to the backup that the score is on. If there is a tie, the score from the most recent backup is chosen.
http://localhost:5000/cal/cs61a/sp16/
http://localhost:5000/cal/cs61a/sp16/hog/
http://localhost:5000/admin/course/1/nikkid%40gmail.com

### CODE REVIEW

**Comment by Sumukh** [comment]
Currently working on separating out the score column for students 

image
image

Mobile:
 
image

**Comment by Sumukh** [comment]
How important is getting the latest backup for breaking ties? The SQLAlchemy query is just a line or two shorter if we can use the latest score instead.

File: server/models.py [hunk 1]
```diff
@@ -599,44 +599,23 @@ def revision(self, user_ids):
...
-                for i, user_id in enumerate(user_ids)
-        }
-        params = {
-            'assignment_id': self.id,
-            **submitter_id_params,
-        }
-
-        score_kinds = self.published_scores if only_published else SCORE_KINDS
-        if not score_kinds:
-            return []  # no published scores
```
Comment (knrafto) [inline server/models.py]: We should keep this case:
```SAWarning: The IN-predicate on "score.kind" was invoked with an empty sequence. This results in a contradiction, which nonetheless can be expensive to evaluate.  Consider alternative strategies for improved performance.
  'strategies for improved performance.' % expr)```

File: server/models.py [hunk 2]
```diff
@@ -313,10 +314,9 @@ class Assignment(Model):
     files = db.Column(JsonBlob)  # JSON object mapping filenames to contents
     course = db.relationship("Course", backref="assignments")
 
-
-    UserAssignment = namedtuple('UserAssignment',
-                                ['assignment', 'subm_time', 'group', 'final_subm'])
-
+    user_assignment = namedtuple('UserAssignment',
```
Comment (knrafto) [inline server/models.py]: Why this change?  `namedtuple` returns a class which are generally named CamelCase

Comment (Sumukh) [inline server/models.py]: I just wanted to make QuantifiedCode happy - I'm ambivalent about the naming 

https://www.quantifiedcode.com/app/project/e6ae811f0c244655b5026e0883ffb8f9/diff/b6b6de97aab18bc2f5a731d8d647395dec2ddf00/9b6eba5fae560e577b645209fd8e5a2f0309b26e/issues?groups=code_patterns%3A3ueqPTsn

**Comment by Sumukh** [comment]
Going to leave this here for future reference: 

A way to get a similiar (but haven't verified that it's identical) result with SQLAlchemy
```
        backups = (db.session.query(Backup.id)
               .filter(Backup.assignment_id == self.id,
                       Backup.submitter_id.in_(user_ids))
               .subquery())
        scores = (db.session.query(Score.kind)
                    .filter(Score.kind.in_(published_kinds))
                    .group_by(Score.kind).subquery())
        score_q = (db.session.query(Score.id)
                   .filter(Score.backup_id == backups.c.id,
                           Score.archived == 0,
                           Score.kind == scores.c.kind)
                   .order_by(Score.score, Score.created)
                   .limit(1)
                   .subquery())
        query = (db.session.query(Score).filter(Score.id == score_q.c.id)
                   .group_by(Score.kind))

```
